lib/translator/read.cc

```diff
@@ -0,0 +1,187 @@
+#include "read.h"
+
+#include "absl/types/span.h"
+
+#include "common.h"
+
+namespace translator {
+
+namespace {  // anonymous namespace for helper functions
+
+// Section 5.3
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode BuildPrinfo(uint8_t rdprotect, uint8_t& prinfo) {
+  bool pract;     // Protection Information Action 1 bit
+  uint8_t prchk;  // Protection Information Check 3 bits
+  prinfo = 0;     // Protection Information field 4 bits
+
+  switch (rdprotect) {  // 3 bits
+    case 0b0:
+      pract = 0b1;
+      prchk = 0b111;
+      break;
+    case 0b001:
+    case 0b101:
+      pract = 0b0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0b0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0b0;
+      prchk = 0b000;
+      break;
+    case 0b100:
+      pract = 0b0;
+      prchk = 0b100;
+      break;
+    default:
+      // Should result in SCSI command termination with status: CHECK CONDITION,
+      //  sense key: ILLEGAL REQUEST, additional sense code: LLEGAL FIELD IN CDB
+      DebugLog("RDPROTECT with value %d has no translation to PRINFO",
+               rdprotect);
+      return StatusCode::kInvalidInput;
+  }
+
+  prinfo |= prchk;
+  prinfo |= (pract << 3);
+
+  return StatusCode::kSuccess;
+}
+
+void SetLbaTags(uint32_t eilbrt, uint16_t elbat, uint16_t elbatm,
+                nvme::GenericQueueEntryCmd& nvme_cmd) {
+  // cdw14 expected initial logical block reference
+  nvme_cmd.cdw[4] = eilbrt;
+  // cdw15 bits 15:0 expected logical block application tag
+  nvme_cmd.cdw[5] |= elbat;
+  // cdw15 bits 31:16 expected logical block application tag mask
+  nvme_cmd.cdw[5] |= (elbatm << 16);
+}
+
+// Translates fields common to all Read commands
+// Named Legacy because it is called directly by Read6, an obsolete command
+// lacking fields common to other Read commands
+StatusCode LegacyRead(uint32_t lba, uint16_t transfer_length,
```

Comment:
LBA needs to be 64 bit; Read16 and nvme both use 64 bit values.

Reply:
Read16 also has a 32 bit transfer length field.

Reply:
@eyhhuang How are you handling the case where th transfer length is greater than 2^16 multiple? According to the translation spec, in case the transfer length is longer, we need to issue multiple nvm express commands. Refer to section 3.14

Reply:
This will be the case with commands Read12/16/32

Reply:
Yes I mentioned this in one of the syncs a while ago but Jonathan and I decided that since NVMe has a transfer limit of 16 bits and SCSI sometimes has a limit of 32 bits, it makes sense for us to enforce that the SCSI read commands do not exceed the NVMe device's limits. 
Since we are the SCSI device, we can set our max transfer limits in Block Limits VPD to the NVMe device's max transfer size  (page size * 2^(MDTS)) which will fit within the NVMe protocol limit of 16 bits.

Reply:
Okay that makes sense. I'll document this in one of the docs as I have a similar case in `Write` Command